### PR TITLE
Create ball filter

### PR DESCRIPTION
**Pull request by NFiedler**
Creates a new Kalman filter based ball filter. This solves issue #2

### CODE REVIEW

**Comment by NFiedler** [comment]
The parameters and integration of the stuff into the behavior are not included. this is just about putting some filter there which actually filters the ball in a somewhat reasonable manner.

**Review by SammyRamone — DISMISSED** [review]
lgtm but didnt test

File: bitbots_ball_filter/src/bitbots_ball_filter/ball_filter.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+                             [0.0, 1.0, 0.0, 1.0],
+                             [0.0, 0.0, 1.0, 0.0],
+                             [0.0, 0.0, 0.0, 1.0]])
+        # measurement function
+        self.kf.H = np.array([[1.0, 0.0, 0.0, 0.0],
+                              [0.0, 1.0, 0.0, 0.0]])
+        # multiplying by the initial uncertainty
+        self.kf.P = np.eye(4) * 1000
+        # assigning measurement noise
+        # TODO: paramter
```
Comment (timonegk) [inline bitbots_ball_filter/src/bitbots_ball_filter/ball_filter.py]: There is still a todo here

File: bitbots_ball_filter/package.xml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  <!-- Use build_export_depend for packages you need in order to build against this package: -->
+  <!--   <build_export_depend>message_generation</build_export_depend> -->
+  <!-- Use buildtool_depend for build tool packages: -->
+  <!--   <buildtool_depend>catkin</buildtool_depend> -->
+  <!-- Use exec_depend for packages you need at runtime: -->
+  <!--   <exec_depend>message_runtime</exec_depend> -->
+  <!-- Use test_depend for packages you need only for testing: -->
+  <!--   <test_depend>gtest</test_depend> -->
+  <!-- Use doc_depend for packages you need only for building documentation: -->
+  <!--   <doc_depend>doxygen</doc_depend> -->
```
Comment (timonegk) [inline bitbots_ball_filter/package.xml]: Clean up boilerplate

File: bitbots_ball_filter/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+## Find catkin macros and libraries
+## if COMPONENTS list like find_package(catkin REQUIRED COMPONENTS xyz)
+## is used, also find other catkin packages
+find_package(catkin REQUIRED COMPONENTS
+  geometry_msgs
+  rospy
+)
+
+## System dependencies are found with CMake's conventions
```
Comment (timonegk) [inline bitbots_ball_filter/CMakeLists.txt]: Please remove the lines that are not needed to make the file more readable.

File: bitbots_ball_filter/scripts/ball_sim.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import rospy
+import numpy as np
+
+from geometry_msgs.msg import PoseWithCovarianceStamped
+
+
+class SimBall:
+    def __init__(self):
+        rospy.init_node('ball_sim')
+        rospy.sleep(1)
```
Comment (timonegk) [inline bitbots_ball_filter/scripts/ball_sim.py]: Is this line necessary?

File: bitbots_ball_filter/scripts/ball_sim.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        self.position = np.zeros((2))
+        self.p_pub = rospy.Publisher('position', PoseWithCovarianceStamped, queue_size=1)
+        self.p_err_pub = rospy.Publisher('ball_relative', PoseWithCovarianceStamped, queue_size=1)
+        while not rospy.is_shutdown():
+            self.position += self.velocity * self.dt
+            self.velocity = np.clip(self.velocity + self.gen_acceleration(), -3, 3)
+            p_err = self.position + self.gen_error()
+            self.p_pub.publish(self.gen_msg(self.position[0], self.position[1]))
+            self.p_err_pub.publish(self.gen_msg(p_err[0], p_err[1]))
+            rospy.sleep(self.dt)
```
Comment (timonegk) [inline bitbots_ball_filter/scripts/ball_sim.py]: Use a `rospy.Rate` instead of this

File: bitbots_ball_filter/scripts/ball_sim.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+    def gen_msg(self, x, y):
+        msg = PoseWithCovarianceStamped()
+        msg.header.frame_id = 'world'
+        msg.pose.pose.position.x = x
+        msg.pose.pose.position.y = y
+        msg.pose.pose.orientation.w = 1
+        return msg
+
+SimBall()
```
Comment (timonegk) [inline bitbots_ball_filter/scripts/ball_sim.py]:
```suggestion
if __name__ == '__main__':
    SimBall()
```

File: bitbots_ball_filter/src/bitbots_ball_filter/ball_filter.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+
+    def filter_step(self, event):
+        """"When ball has been assigned a value and filter has been initialized:
+            state will be updated according to filter.
+            If filter has not been initialized, then that will be done first.
+
+            If there is not data for ball, a prediction will still be made and published
+            Process noise is taken into account
+            """
```
Comment (timonegk) [inline bitbots_ball_filter/src/bitbots_ball_filter/ball_filter.py]:
```suggestion
        """"
        When ball has been assigned a value and filter has been initialized:
        state will be updated according to filter.
        If filter has not been initialized, then that will be done first.

        If there is not data for ball, a prediction will still be made and published
        Process noise is taken into account
        """
```

File: bitbots_ball_filter/src/bitbots_ball_filter/ball_filter.py [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+                state = self.kf.get_update()
+                self.filter_init = True
+            self.publish_data(*state)
+            self.ball = None
+        else: 
+            if self.filter_init:
+                self.kf.predict()
+                self.kf.update(None)
+                state = self.kf.get_update()
+                self.publish_data(*state)
```
Comment (timonegk) [inline bitbots_ball_filter/src/bitbots_ball_filter/ball_filter.py:108]: In the if-branches above, you could move common parts outside of the branches.

File: bitbots_ball_filter/src/bitbots_ball_filter/ball_filter.py [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+        :param y: start y position of the ball
+        :return:
+        """
+        # initial value of position(x,y) of the ball and velocity
+        self.kf.x = np.array([x, y, 0, 0])
+        # transition matrix
+        self.kf.F = np.array([[1.0, 0.0, 1.0, 0.0],
+                             [0.0, 1.0, 0.0, 1.0],
+                             [0.0, 0.0, 1.0, 0.0],
+                             [0.0, 0.0, 0.0, 1.0]])
```
Comment (timonegk) [inline bitbots_ball_filter/src/bitbots_ball_filter/ball_filter.py:129]: Where do these values come from?

**Review by timonegk — CHANGES_REQUESTED** [review]

**Review by Flova — DISMISSED** [review]
I tested this pr and it works

**Review by timonegk — APPROVED** [review]

**Pull request merged**